elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala

```diff
@@ -148,9 +148,15 @@ class RestlasticSearchClient(endpointProvider: EndpointProvider, signer: Option[
   }
 
   // retryOnConflictOpt specifies how many times to retry before throwing version conflict exception.
-  // https://www.elastic.co/guide/en/elasticsearch/reference/2.3/docs-update.html#_parameters_2
-  def bulkUpdate(index: Index, tpe: Type, documents: Seq[Document], retryOnConflictOpt: Option[Int] = None): Future[Seq[BulkItem]] = {
-    val bulkOperation = Bulk(documents.map(BulkOperation(update, Some(index -> tpe), _, retryOnConflictOpt)))
+  // upsertOpt specifies the document to create in an update request when the document does not exist.
+  // https://www.elastic.co/guide/en/elasticsearch/reference/2.3/docs-update.html
+  def bulkUpdate(index: Index, tpe: Type, documents: Seq[Document], retryOnConflictOpt: Option[Int] = None, upsertsOpt: Option[Seq[Document]] = None): Future[Seq[BulkItem]] = {
```

Comment:
This interface seems fragile because it insists on `documents` being zippable with `upsertsOpt`.

I guess I do not understand why if I want to do an upsert, I have to specify it in 2 places, one in `documents`, and another in `upsertsOpt`?

Reply:
It seems we are trying to 'bulk'-ify this particular api: https://www.elastic.co/guide/en/elasticsearch/reference/2.3/docs-update.html#_literal_doc_as_upsert_literal

Would `bulkUpdate` that, instead of `upsertsOpt`,  takes a boolean flag `isUpsert`, and applies it to all `documents`, work for our use case?
```
def bulkUpdate(index: Index, tpe: Type, documents: Seq[Document], retryOnConflictOpt: Option[Int] = None, isUpsert: Boolean): Future[Seq[BulkItem]]
```

Reply:
Oh, this is the old behavior where the whole document is specified for update and upsert. As a result, we have to know the entire content of the document. 

The new behavior is trying to support the use case of:
1. We have first_seen_timstamp in the document
2. When we do the update call we want to 
    a. create the document with the timestamp provided if the document does not exist
    b. maintain the old first_seen_timestamp, but update the rest of the document

so the document will be something like "cluster=kafka, metrics=cpu", and the contents in upsert will be "cluster=kafka, metrics=cpu, first_seen_timestamp=124238993"

Reply:
Well, I might be wrong, but it feels like you could mix bulk operations when querying bulk endpoint. Then we should have interface which looks more or less like this:

> 
> trait Operation {
> def toJson()
> // other needed methods
> }
> case class Update(document: Document) extends Operation
> case class Insert(document: Document)extends Operation
> case class Delete(document: Document)extends Operation
> case class Upsert(document: Document, upsert: Option[Document], script: Option[??]) extends Operation {
> def this(document: Document) = this(document, None, None)
> }
> 
> def bulk(index: Index, tpe: Type, operations: Seq[Operation], retryOnConflictOpt: Option[Int] = None): Future[Seq[BulkItem]]
>

wdyt?

Reply:
Whats more, you could differentiate between:
Upsert with doc_as_upsert -> true
Upsert with upsert-> document_data
and have separate functions for bulking each of them.